### PR TITLE
feat(config): #13 - config handling using koanf package

### PR DESCRIPTION
**Pull request by DOO-DEV**

### CODE REVIEW

File: config/config_test.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
+package config
```
Comment (gohossein) [inline config/config_test.go:1]: Please change `config` to `config_test`.
All test files should have `_test` postfix in the package name.

**Review by gohossein — COMMENTED** [review]
Add `test` command to `Makefile`.
Add `Github Action` to run `make test` command in `pull request` and `main` branch.

**Review by gohossein — COMMENTED** [review]
Add `ormus` scope to PR title
`feat(ormus): config....`

File: Makefile [hunk 2]
```diff
@@ -8,4 +8,7 @@ ROOT=$(realpath $(dir $(lastword $(MAKEFILE_LIST))))
 // TODO: add lint and format to github ci
 lint:
 	which golangci-lint || (go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.54.0)
-	golangci-lint run --config=$(ROOT)/.golangci.yml $(ROOT)/...
\ No newline at end of file
+	golangci-lint run --config=$(ROOT)/.golangci.yml $(ROOT)/...
```
Comment (gohossein) [inline Makefile:11]: You need to define `ROOT` variable.

File: Makefile [hunk 3]
```diff
@@ -8,4 +8,7 @@ ROOT=$(realpath $(dir $(lastword $(MAKEFILE_LIST))))
 // TODO: add lint and format to github ci
```
Comment (gohossein) [inline Makefile:8]: Also add `format` command and remove `TODO` comment

File: Makefile [hunk 4]
```diff
@@ -8,4 +8,7 @@ ROOT=$(realpath $(dir $(lastword $(MAKEFILE_LIST))))
...
 lint:
 	which golangci-lint || (go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.54.0)
-	golangci-lint run --config=$(ROOT)/.golangci.yml $(ROOT)/...
\ No newline at end of file
+	golangci-lint run --config=$(ROOT)/.golangci.yml $(ROOT)/...
+
+test:
+	go test -v ./...
```
Comment (gohossein) [inline Makefile:14]: I guess the indentation with `4 spaces` is enough.

**Review by gohossein — COMMENTED** [review]
Looks great to me, just a few things need to be fixed.

File: Makefile (hunk 3 above)
Comment (DOO-DEV) [reply Makefile:8]: the format command add to the #16 pull request

File: Makefile (hunk 2 above)
Comment (DOO-DEV) [reply Makefile:11]: `Root` is already defined

File: Makefile (hunk 4 above)
Comment (DOO-DEV) [reply Makefile:14]: I use one tab and its 4 space. in the `main`  branch that you write it its the same spacing

image

**Review by gohossein — APPROVED** [review]

**Comment by gohossein** [comment]
Please `rebase` with `main` branch and resovle the conflicts.

**Comment by gohossein** [comment]
Please add issue number to PR title like [these](https://github.com/ormushq/ormus/pull/17) PR.

File: config/loader.go [hunk 5]
```diff
@@ -1,56 +1,85 @@
...
-	"fmt"
-	"log"
-	"strings"
-
 	"github.com/knadh/koanf"
 	"github.com/knadh/koanf/parsers/yaml"
 	"github.com/knadh/koanf/providers/env"
 	"github.com/knadh/koanf/providers/file"
 	"github.com/knadh/koanf/providers/structs"
+	"log"
```
Comment (gohossein) [inline config/loader.go:9]: sort import packages using `goimports`, see [this](https://bariseser.medium.com/useful-go-command-1-goimports-7a9eaa7869a3).

File: config/loader.go [hunk 6]
```diff
@@ -1,56 +1,85 @@
...
 	}
 
-	var instance Config
-	if err := k.Unmarshal("", &instance); err != nil {
+	if err := k.Unmarshal("", &c); err != nil {
 		log.Fatalf("error unmarshaling config: %s", err)
 	}
+}
+
+func C() *Config {
```
Comment (gohossein) [inline config/loader.go:61]: Do not expose Config with pointer. The caller can modify the config object, just return `Config`.

File: config/loader.go [hunk 7]
```diff
@@ -1,56 +1,85 @@
...
+	return c
+}
+
+func New(opt Option) *Config {
```
Comment (gohossein) [inline config/loader.go:65]: Just return `Config`.

**Review by gohossein — COMMENTED** [review]
Very close to Approve :)

File: config/loader.go (hunk 6 above)
Comment (DOO-DEV) [reply config/loader.go]: done

File: config/loader.go (hunk 7 above)
Comment (DOO-DEV) [reply config/loader.go]: done

**Review by gohossein — APPROVED** [review]
Great

**Pull request merged**